File: movecraft-api/src/main/java/net.countercraft.movecraft.api/MaterialDataPredicate.java
package net.countercraft.movecraft.api;

import com.google.common.base.Joiner;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Iterables;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.material.MaterialData;

import java.util.Collections;
import java.util.EnumSet;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

public abstract class MaterialDataPredicate {
    private MaterialDataPredicate() {
    }

    public abstract boolean isTrivial();

    public abstract boolean check(MaterialData materialData);
    public boolean check(Material material) {
        return check(new MaterialData(material));
    }
    public boolean check(Material material, byte data) {
        return check(new MaterialData(material, data));
    }
    public boolean checkBlock(Block block) {
        return check(block.getState().getData());
    }

    public abstract Iterable<Material> allMaterials();
    public abstract Iterable<MaterialData> allMaterialDataPairs();

    private static final None NONE = new None();
    private static final AllBlocks ALL_BLOCKS = new AllBlocks();

    public static MaterialDataPredicate none() {
        return NONE;
    }

    public static MaterialDataPredicate all() {
        return ALL_BLOCKS;
    }

    public static MaterialDataPredicate single(Material material) {
        return new SingleMaterial(material);
    }

    public static MaterialDataPredicate single(MaterialData materialData) {
        return new SingleMaterialData(materialData);
    }

    public static MaterialDataPredicate single(Material material, byte data) {
        return new SingleMaterialData(new MaterialData(material, data));
    }

    public static MaterialDataPredicate many(Set<Material> materials, Set<MaterialData> materialDataPairs) {
        if (materials.isEmpty() && materialDataPairs.isEmpty()) {
            return NONE;
        } else if (materials.size() == 1 && materialDataPairs.isEmpty()) {
            return new SingleMaterial(materials.toArray(new Material[1])[0]);
        } else if (materialDataPairs.size() == 1 && materials.isEmpty()) {
            return new SingleMaterialData(materialDataPairs.toArray(new MaterialData[1])[0]);
        } else {
            return new Many(materials, materialDataPairs);
        }
    }

    private static final class None extends MaterialDataPredicate {
        @Override public boolean isTrivial() {
            return true;
        }

        @Override public boolean check(MaterialData materialData) {
            return false;
        }

        @Override public Iterable<Material> allMaterials() {
            return Collections.emptyList();
        }

        @Override public Iterable<MaterialData> allMaterialDataPairs() {
            return Collections.emptyList();
        }

        @Override public String toString() {
            return "()";
        }

        @Override public int hashCode() {
            return NONE.hashCode();
        }

        @Override public boolean equals(Object o) {
            return NONE.equals(o);
        }
    }

    private static final class AllBlocks extends MaterialDataPredicate {
        private static final ImmutableSet<Material> NON_AIR_BLOCK_MATERIALS;
        static {
            ImmutableSet.Builder<Material> builder = ImmutableSet.builder();
            for (Material material : EnumSet.allOf(Material.class)) {
                if (material != Material.AIR && material.isBlock()) {
                    builder.add(material);
                }
            }
            NON_AIR_BLOCK_MATERIALS = builder.build();
        }

        @Override public boolean isTrivial() {
            return !NON_AIR_BLOCK_MATERIALS.isEmpty();
        }

        @Override public boolean check(MaterialData materialData) {
            return materialData.getItemType() != Material.AIR && materialData.getItemType().isBlock();
        }

        @Override public Iterable<Material> allMaterials() {
            return NON_AIR_BLOCK_MATERIALS;
        }

        @Override public Iterable<MaterialData> allMaterialDataPairs() {
            return Collections.emptyList();
        }

        @Override public String toString() {
            return "(*)";
        }

        @Override public int hashCode() {
            return ALL_BLOCKS.hashCode();
        }

        @Override public boolean equals(Object o) {
            return ALL_BLOCKS.equals(o);
        }
    }

    private static final class SingleMaterial extends MaterialDataPredicate {
        public final Material material;

        public SingleMaterial(Material material) {
            Preconditions.checkNotNull(material);
            this.material = material;
        }

        @Override public boolean isTrivial() {
            return false;
        }

        @Override public boolean check(MaterialData materialData) {
            return material == materialData.getItemType();
        }

        @Override public Iterable<Material> allMaterials() {
            return ImmutableList.of(material);
        }

        @Override public Iterable<MaterialData> allMaterialDataPairs() {
            return Collections.emptyList();
        }

        @Override public int hashCode() {
            return material.hashCode();
        }

        @Override public String toString() {
            return "(" + material + ')';
        }

        @Override public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            SingleMaterial that = (SingleMaterial) o;
            return material == that.material;
        }
    }

    private static final class SingleMaterialData extends MaterialDataPredicate {
        public final MaterialData materialData;

        public SingleMaterialData(MaterialData materialData) {
            Preconditions.checkNotNull(materialData);
            this.materialData = materialData;
        }

        @Override public boolean isTrivial() {
            return false;
        }

        @Override public boolean check(MaterialData materialData) {
            return Objects.equals(this.materialData, materialData);
        }

        @Override public Iterable<Material> allMaterials() {
            return Collections.emptyList();
        }

        @Override public Iterable<MaterialData> allMaterialDataPairs() {
            return ImmutableList.of(materialData);
        }

        @Override public int hashCode() {
            return materialData.hashCode();
        }

        @Override public String toString() {
            return "(" + materialData + ')';
        }

        @Override public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            SingleMaterialData that = (SingleMaterialData) o;

            return materialData.equals(that.materialData);
        }
    }

    private static final class Many extends MaterialDataPredicate {
        public final ImmutableSet<Material> materials;
        public final ImmutableSet<MaterialData> materialDataPairs;

        public Many(Set<Material> materials, Set<MaterialData> materialDataPairs) {
            Preconditions.checkNotNull(materials);
            Preconditions.checkNotNull(materialDataPairs);
            Preconditions.checkArgument(materials.size() + materialDataPairs.size() > 1,
                                        "Use SingleMaterial or SingleMaterialData for predicates with a single " +
                                        "element.");
            this.materials = ImmutableSet.copyOf(materials);
            this.materialDataPairs = ImmutableSet.copyOf(materialDataPairs);
        }

        @Override public boolean isTrivial() {
            return false;
        }

        @Override public boolean check(MaterialData materialData) {
            return materials.contains(materialData.getItemType()) || materialDataPairs.contains(materialData);
        }

        @Override public Iterable<Material> allMaterials() {
            return materials;
        }

        @Override public Iterable<MaterialData> allMaterialDataPairs() {
            return materialDataPairs;
        }

        @Override public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Many that = (Many) o;

            return materials.equals(that.materials) && materialDataPairs.equals(that.materialDataPairs);
        }

        @Override public int hashCode() {
            int result = materials.hashCode();
            result = 31 * result + materialDataPairs.hashCode();
            return result;
        }

        private static final int AVERAGE_NAME_LENGTH = 10;
        private static final Joiner joiner = Joiner.on(" | ");

        @Override public String toString() {
            StringBuilder result = new StringBuilder(
                    AVERAGE_NAME_LENGTH * (materials.size() + materialDataPairs.size()));

            result.append('(');
            joiner.appendTo(result, Iterables.concat(materials, materialDataPairs));
            result.append(')');

            return result.toString();
        }
    }

    public static class Builder {
        private final EnumSet<Material> materials;
        private final HashSet<MaterialData> materialDataPairs;
        private boolean all;

        public Builder() {
            this.materials = EnumSet.noneOf(Material.class);
            this.materialDataPairs = new HashSet<>();
            this.all = false;
        }

        public void add(Material material) {
            if (!all) materials.add(material);
        }

        public void add(MaterialData materialDataPair) {
            if (!all) materialDataPairs.add(materialDataPair);
        }

        public void add(MaterialDataPredicate predicate) {
            if (!all) {
                if (predicate instanceof SingleMaterial) {
                    materials.add(((SingleMaterial) predicate).material);
                } else if (predicate instanceof SingleMaterialData) {
                    materialDataPairs.add(((SingleMaterialData) predicate).materialData);
                } else if (predicate instanceof Many) {
                    materials.addAll(((Many) predicate).materials);
                    materialDataPairs.addAll(((Many) predicate).materialDataPairs);
                } else if (predicate instanceof AllBlocks) {
                    all = true;
                    materials.clear();
                    materialDataPairs.clear();
                }
            }
        }

        public MaterialDataPredicate result() {
            if (all) return all();
            else return many(materials, materialDataPairs);
        }
    }
}

File: movecraft-api/src/main/scala/com/alexknvl/shipcraft/MaterialDataPredicate.scala
package com.alexknvl.shipcraft

import com.google.common.base.Preconditions
import org.bukkit.Material
import org.bukkit.block.Block
import org.bukkit.material.MaterialData

sealed abstract class MaterialDataPredicate {
  def isTrivial: Boolean

  def check(materialData: MaterialData): Boolean
  def check(material: Material): Boolean = check(new MaterialData(material))
  def check(material: Material, data: Byte): Boolean = check(new MaterialData(material, data))
  def checkBlock(block: Block): Boolean = check(block.getState.getData)

  def allMaterials: Set[Material]
  def allMaterialDataPairs: Set[MaterialData]
}

object MaterialDataPredicate {
  def none: MaterialDataPredicate = None

  def all: MaterialDataPredicate = AllBlocks

  def single(material: Material): MaterialDataPredicate =
    SingleMaterial(material)

  def single(materialData: MaterialData): MaterialDataPredicate =
    SingleMaterialData(materialData)

  def single(material: Material, data: Byte): MaterialDataPredicate =
    SingleMaterialData(new MaterialData(material, data))

  def many(materials: Set[Material], materialDataPairs: Set[MaterialData]): MaterialDataPredicate =
    if (materials.isEmpty && materialDataPairs.isEmpty) None
    else if (materials.size == 1 && materialDataPairs.isEmpty) SingleMaterial(materials.head)
    else if (materialDataPairs.size == 1 && materials.isEmpty) SingleMaterialData(materialDataPairs.head)
    else Many(materials, materialDataPairs)

  final private case object None extends MaterialDataPredicate {
    def isTrivial = true

    def check(materialData: MaterialData) = false

    def allMaterials: Set[Material] = Set.empty

    def allMaterialDataPairs: Set[MaterialData] = Set.empty

    override def toString = "()"
  }

  final private case object AllBlocks extends MaterialDataPredicate {
    import scala.collection.JavaConverters._
    private val NON_AIR_BLOCK_MATERIALS = for {
      material <- java.util.EnumSet.allOf(classOf[Material]).asScala.toSet
      if material.isBlock
      if material ne Material.AIR
    } yield material

    def isTrivial: Boolean = NON_AIR_BLOCK_MATERIALS.nonEmpty

    def check(materialData: MaterialData): Boolean = (materialData.getItemType ne Material.AIR) && materialData.getItemType.isBlock

    def allMaterials = NON_AIR_BLOCK_MATERIALS

    def allMaterialDataPairs: Set[MaterialData] = Set.empty

    override def toString = "(*)"
  }

  final private case class SingleMaterial(material: Material) extends MaterialDataPredicate {
    Preconditions.checkNotNull(material)

    def isTrivial = false

    def check(materialData: MaterialData): Boolean = material eq materialData.getItemType

    def allMaterials: Set[Material] = Set(material)

    def allMaterialDataPairs: Set[MaterialData] = Set.empty

    override def toString: String = "(" + material + ')'
  }

  final private case class SingleMaterialData(materialData: MaterialData) extends MaterialDataPredicate {
    Preconditions.checkNotNull(materialData)

    def isTrivial = false

    def check(materialData: MaterialData): Boolean = this.materialData == materialData

    def allMaterials: Set[Material] = Set.empty

    def allMaterialDataPairs: Set[MaterialData] = Set(materialData)

    override def toString: String = "(" + materialData + ')'
  }

  final private case class Many(materials: Set[Material], materialDataPairs: Set[MaterialData]) extends MaterialDataPredicate {
    require(materials.size + materialDataPairs.size > 1)

    def isTrivial = false

    def check(materialData: MaterialData): Boolean = materials.contains(materialData.getItemType) || materialDataPairs.contains(materialData)

    def allMaterials: Set[Material] = materials

    def allMaterialDataPairs: Set[MaterialData] = materialDataPairs

    override def toString: String =
      "(" + (materials.map(_.toString) ++ materialDataPairs.map(_.toString)).mkString(" | ") + ")"
  }

  class Builder() {
    final private val materials = Set.newBuilder[Material]
    final private val materialDataPairs = Set.newBuilder[MaterialData]
    private var all = false

    def add(material: Material): Unit = if (!all) materials.+=(material)

    def add(materialDataPair: MaterialData): Unit = if (!all) materialDataPairs.+=(materialDataPair)

    def add(predicate: MaterialDataPredicate): Unit =
      if (!all) predicate match {
        case MaterialDataPredicate.None => ()
        case MaterialDataPredicate.SingleMaterial(x) => materials.+=(x)
        case MaterialDataPredicate.SingleMaterialData(x) => materialDataPairs.+=(x)
        case MaterialDataPredicate.Many(a, b) =>
          a.foreach(x => materials.+=(x))
          b.foreach(x => materialDataPairs.+=(x))
        case AllBlocks =>
          all = true
          materials.clear()
          materialDataPairs.clear()
      }

    def result: MaterialDataPredicate =
      if (all) AllBlocks
      else many(materials.result(), materialDataPairs.result())
  }

}